checker implemented

jennatnguyen opened the pull request.

kelloggm commented on README.md:
```diff
@@ -7,74 +7,4 @@ in this directory, or on your course webpage.
 The remainder of this README is the user-facing documentation for the analysis
 you will build.
 
-You should remove, from your version, everything above and including this line.
```
I'm surprised that you deleted this and everything _below_ it, rather than _above_. Was there a particular reason?

kelloggm commented on src/main/java/org/checkerframework/checker/dividebyzero/DivByZeroTransfer.java:
```diff
@@ -74,11 +74,39 @@ enum BinaryOperator {
+    } else if (operator == Comparison.LT) {
+        // If lhs < 0, we can say lhs is less than Zero
+        if (equal(lhs, reflect(Zero.class))) {
+            return reflect(Zero.class); // Refine lhs to Zero
```
This doesn't seem correct - I would expect NonZero here.

kelloggm commented on src/main/java/org/checkerframework/checker/dividebyzero/DivByZeroTransfer.java:
```diff
@@ -74,11 +74,39 @@ enum BinaryOperator {
+        }
+    } else if (operator == Comparison.LE) {
+        if (equal(rhs, reflect(Zero.class))) {
+            return reflect(Zero.class); // lhs could be zero if less than or equal to Zero
```
This is surprising (and seems unsound to me): it's implying that if I have the following code:
```
if (x <= 0) { ... }
```
Then, inside the `if` block, `x` must be zero. I think it _might_ be zero, but I don't think it's guaranteed to be zero. I suggest trying to write an example where this unsoundness leads to a missed division by zero error; as a hint, you probably want to compare `x` against some other variable, and then divide by that one.

kelloggm replied:
(The same problem exists for `>=`)

kelloggm commented on src/main/java/org/checkerframework/checker/dividebyzero/DivByZeroTransfer.java:
```diff
@@ -95,11 +123,55 @@ private AnnotationMirror refineLhsOfComparison(
+    if (operator == BinaryOperator.DIVIDE) {
+        // Check for division by zero
+        if (equal(rhs, reflect(Zero.class)) || equal(lhs, top())) {
+            return reflect(Zero.class); // Division by zero detected
```
I would expect this to be bottom or nonzero rather than zero, since it implies a division by zero error (in the error case) or that the value was actually nonzero (if the program can continue to execute).

kelloggm commented on src/main/java/org/checkerframework/checker/dividebyzero/DivByZeroTransfer.java:
```diff
@@ -95,11 +123,55 @@ private AnnotationMirror refineLhsOfComparison(
+    // Handle addition
+    if (operator == BinaryOperator.PLUS) {
+        if (!equal(lhs, reflect(Zero.class)) && !equal(rhs, reflect(Zero.class))) {
+            return reflect(NonZero.class); // Both operands are NonZero
```
Is this sound? I think it gives the wrong answer for an expression like `-3 + 3`.

kelloggm commented on src/main/java/org/checkerframework/checker/dividebyzero/DivByZeroTransfer.java:
```diff
@@ -95,11 +123,55 @@ private AnnotationMirror refineLhsOfComparison(
+        if (!equal(lhs, reflect(Zero.class)) && !equal(rhs, reflect(Zero.class))) {
+            return reflect(NonZero.class); // Both operands are NonZero
+        }
+        return lub(lhs, rhs); // Return least upper bound
```
Why least upper bound here?

kelloggm commented on src/main/java/org/checkerframework/checker/dividebyzero/DivByZeroVisitor.java:
```diff
@@ -29,7 +29,18 @@ public class DivByZeroVisitor extends BaseTypeVisitor<DivByZeroAnnotatedTypeFact
+
+	// Check if the right-hand operand could be zero
+	if (rhs != null && hasAnnotation(rhs, Zero.class)) {
+	    return true;  // Only flag a divide-by-zero if rhs has @Zero
```
I don't think this is sound, since `top` can logically also contain zero, right? Top represents "we don't know anything" - which can mean that the value is either zero or nonzero. So, to guarantee safety, the analysis needs to warn about top, too.

jennatnguyen commented:
In order to handle cancellation I added Positive and Negative to my
lattice. A lot of my issues were fixed by adding error checks at top(),
thanks for that tidbit,

Updated code:
https://github.com/jennatnguyen/div-by-zero-checker/pull/3/files

On Tue, Oct 15, 2024 at 8:49 AM Martin Kellogg ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In README.md
> <https://github.com/jennatnguyen/div-by-zero-checker/pull/2#discussion_r1801091431>
> :
>
> > @@ -7,74 +7,4 @@ in this directory, or on your course webpage.
>  The remainder of this README is the user-facing documentation for the analysis
>  you will build.
>
> -You should remove, from your version, everything above and including this line.
>
> I'm surprised that you deleted this and everything *below* it, rather
> than *above*. Was there a particular reason?
> ------------------------------
>
> In
> src/main/java/org/checkerframework/checker/dividebyzero/DivByZeroTransfer.java
> <https://github.com/jennatnguyen/div-by-zero-checker/pull/2#discussion_r1801093096>
> :
>
> > +
> +private AnnotationMirror refineLhsOfComparison(
> +    Comparison operator, AnnotationMirror lhs, AnnotationMirror rhs) {
> +
> +    if (operator == Comparison.NE) {
> +        if (equal(rhs, reflect(Zero.class))) {
> +            return reflect(NonZero.class); // lhs cannot be Zero if it is not equal to Zero
> +        }
> +    } else if (operator == Comparison.EQ) {
> +        if (equal(rhs, reflect(Zero.class))) {
> +            return reflect(Zero.class); // Refine lhs to Zero
> +        }
> +    } else if (operator == Comparison.LT) {
> +        // If lhs < 0, we can say lhs is less than Zero
> +        if (equal(lhs, reflect(Zero.class))) {
> +            return reflect(Zero.class); // Refine lhs to Zero
>
> This doesn't seem correct - I would expect NonZero here.
> ------------------------------
>
> In
> src/main/java/org/checkerframework/checker/dividebyzero/DivByZeroTransfer.java
> <https://github.com/jennatnguyen/div-by-zero-checker/pull/2#discussion_r1801095997>
> :
>
> > +    if (operator == Comparison.NE) {
> +        if (equal(rhs, reflect(Zero.class))) {
> +            return reflect(NonZero.class); // lhs cannot be Zero if it is not equal to Zero
> +        }
> +    } else if (operator == Comparison.EQ) {
> +        if (equal(rhs, reflect(Zero.class))) {
> +            return reflect(Zero.class); // Refine lhs to Zero
> +        }
> +    } else if (operator == Comparison.LT) {
> +        // If lhs < 0, we can say lhs is less than Zero
> +        if (equal(lhs, reflect(Zero.class))) {
> +            return reflect(Zero.class); // Refine lhs to Zero
> +        }
> +    } else if (operator == Comparison.LE) {
> +        if (equal(rhs, reflect(Zero.class))) {
> +            return reflect(Zero.class); // lhs could be zero if less than or equal to Zero
>
> This is surprising (and seems unsound to me): it's implying that if I have
> the following code:
>
> if (x <= 0) { ... }
>
> Then, inside the if block, x must be zero. I think it *might* be zero,
> but I don't think it's guaranteed to be zero. I suggest trying to write an
> example where this unsoundness leads to a missed division by zero error; as
> a hint, you probably want to compare x against some other variable, and
> then divide by that one.
> ------------------------------
>
> In
> src/main/java/org/checkerframework/checker/dividebyzero/DivByZeroTransfer.java
> <https://github.com/jennatnguyen/div-by-zero-checker/pull/2#discussion_r1801097988>
> :
>
> > @@ -95,11 +123,55 @@ private AnnotationMirror refineLhsOfComparison(
>     * @param rhs the lattice point for the right-hand side of the expression
>     * @return the lattice point for the result of the expression
>     */
> -  private AnnotationMirror arithmeticTransfer(
> -      BinaryOperator operator, AnnotationMirror lhs, AnnotationMirror rhs) {
> -    // TODO
> +private AnnotationMirror arithmeticTransfer(
> +    BinaryOperator operator, AnnotationMirror lhs, AnnotationMirror rhs) {
> +
> +    if (operator == BinaryOperator.DIVIDE) {
> +        // Check for division by zero
> +        if (equal(rhs, reflect(Zero.class)) || equal(lhs, top())) {
> +            return reflect(Zero.class); // Division by zero detected
>
> I would expect this to be bottom or nonzero rather than zero, since it
> implies a division by zero error (in the error case) or that the value was
> actually nonzero (if the program can continue to execute).
> ------------------------------
>
> In
> src/main/java/org/checkerframework/checker/dividebyzero/DivByZeroTransfer.java
> <https://github.com/jennatnguyen/div-by-zero-checker/pull/2#discussion_r1801098895>
> :
>
> > +            return reflect(NonZero.class); // Result is NonZero
> +        }
> +
> +        // If lhs is Zero, the result of division is Zero
> +        if (equal(lhs, reflect(Zero.class))) {
> +            return reflect(Zero.class);
> +        }
> +
> +        // If both lhs and rhs are unknown, we can't make any conclusions
> +        return top();
> +    }
> +
> +    // Handle addition
> +    if (operator == BinaryOperator.PLUS) {
> +        if (!equal(lhs, reflect(Zero.class)) && !equal(rhs, reflect(Zero.class))) {
> +            return reflect(NonZero.class); // Both operands are NonZero
>
> Is this sound? I think it gives the wrong answer for an expression like -3
> + 3.
> ------------------------------
>
> In
> src/main/java/org/checkerframework/checker/dividebyzero/DivByZeroTransfer.java
> <https://github.com/jennatnguyen/div-by-zero-checker/pull/2#discussion_r1801099340>
> :
>
> > +
> +        // If lhs is Zero, the result of division is Zero
> +        if (equal(lhs, reflect(Zero.class))) {
> +            return reflect(Zero.class);
> +        }
> +
> +        // If both lhs and rhs are unknown, we can't make any conclusions
> +        return top();
> +    }
> +
> +    // Handle addition
> +    if (operator == BinaryOperator.PLUS) {
> +        if (!equal(lhs, reflect(Zero.class)) && !equal(rhs, reflect(Zero.class))) {
> +            return reflect(NonZero.class); // Both operands are NonZero
> +        }
> +        return lub(lhs, rhs); // Return least upper bound
>
> Why least upper bound here?
> ------------------------------
>
> In
> src/main/java/org/checkerframework/checker/dividebyzero/DivByZeroVisitor.java
> <https://github.com/jennatnguyen/div-by-zero-checker/pull/2#discussion_r1801102150>
> :
>
> > @@ -29,7 +29,18 @@ public class DivByZeroVisitor extends BaseTypeVisitor<DivByZeroAnnotatedTypeFact
>    private boolean errorAt(BinaryTree node) {
>      // A BinaryTree can represent any binary operator, including + or -.
>      // TODO
> -    return false;
> +
> +    // Check if the operator is one of the division or remainder operators
> +	if (DIVISION_OPERATORS.contains(node.getKind())) {
> +	// Get the right-hand side operand
> +	Tree rhs = node.getRightOperand();
> +
> +	// Check if the right-hand operand could be zero
> +	if (rhs != null && hasAnnotation(rhs, Zero.class)) {
> +	    return true;  // Only flag a divide-by-zero if rhs has @Zero
>
> I don't think this is sound, since top can logically also contain zero,
> right? Top represents "we don't know anything" - which can mean that the
> value is either zero or nonzero. So, to guarantee safety, the analysis
> needs to warn about top, too.
> ------------------------------
>
> In
> src/main/java/org/checkerframework/checker/dividebyzero/DivByZeroTransfer.java
> <https://github.com/jennatnguyen/div-by-zero-checker/pull/2#discussion_r1801096444>
> :
>
> > +    if (operator == Comparison.NE) {
> +        if (equal(rhs, reflect(Zero.class))) {
> +            return reflect(NonZero.class); // lhs cannot be Zero if it is not equal to Zero
> +        }
> +    } else if (operator == Comparison.EQ) {
> +        if (equal(rhs, reflect(Zero.class))) {
> +            return reflect(Zero.class); // Refine lhs to Zero
> +        }
> +    } else if (operator == Comparison.LT) {
> +        // If lhs < 0, we can say lhs is less than Zero
> +        if (equal(lhs, reflect(Zero.class))) {
> +            return reflect(Zero.class); // Refine lhs to Zero
> +        }
> +    } else if (operator == Comparison.LE) {
> +        if (equal(rhs, reflect(Zero.class))) {
> +            return reflect(Zero.class); // lhs could be zero if less than or equal to Zero
>
> (The same problem exists for >=)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jennatnguyen/div-by-zero-checker/pull/2#pullrequestreview-2369219716>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BCWIO7WFCGQBA5K5KTJF5SLZ3UFNXAVCNFSM6AAAAABP5WUCF2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDGNRZGIYTSNZRGY>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>
